Support NFSv4 only environment

benjamreis opened the pull request:
NFSv4 only environments don't support `rpcinfo`
and `showmount` calls as they're missing NFSv3
services.

When `rpcinfo` or `showmount` fails, try to mount
NFSv4 pseudo FS '/' to add '4' to supported NFS versions
and run `ls` on the mounted pseudo FS to offer the first
folder level of exports.

See: https://github.com/xapi-project/sm/issues/551
And: https://github.com/xcp-ng/xcp/issues/135

benjamreis commented:
What can be done next:

Attempt a mount on / with nfs4 and:
- run `nfsstat -m` to get the version of the mount and add it to the supported version
- run ls in the mounted / to get some exported paths BUT what if there's multiple level of folders?

What do you think @MarkSymsCtx?

MarkSymsCtx commented:
> What can be done next:
> 
> Attempt a mount on / with nfs4 and:
> 
> * run `nfsstat -m` to get the version of the mount and add it to the supported version
> * run ls in the mounted / to get some exported paths BUT what if there's multiple level of folders?
> 
> What do you think @MarkSymsCtx?

It's all generally horrible with NFSv4 but the above may help in a significant set of sub-cases.

benjamreis commented:
Okay, i'll try that then, it's still better than nothing.
Would you prefer to merge this PR 1st and I'll do a 2nd one on top or wait and do everything in this PR?

stormi commented:
> run nfsstat -m to get the version of the mount and add it to the supported version 

I don't know exactly how this works, but the current version of `mount.nfs` in XS/XCP-ng dom0 defaulted to 4.1 for a server that only supports 4.1 and 4.2. Another client on a more recent computer defaulted to 4.2 for the same server.

This means that the version we can extract from `nfsstat -m` depends on the client and doesn't give much information about what the server supports, so maybe `probe` should just reply `4` as the supported version when there's no NFS v3 (Option A). 

Another approach would be, after a first successful mount of `/`, to attempt to mount `/` with every known version of the v4 protocol (4.0, 4.1, 4.2), and keep only the working values (Option B).

Yet another approach would be always listing all version of NFSv4, and let it fail at `sr-create` time when the user attempts a not-supported version (Option C).

stormi commented on drivers/nfs.py:
```diff
@@ -282,11 +318,26 @@ def get_supported_nfs_versions(server):
+    except Exception:
+        util.SMlog("Unable to obtain list of valid nfs versions with rpcinfo, trying NSFv4")
+
+    try:
+        mountpoint = "%s/%s" % NFS4_TMP_MOUNTPOINT, server
+        soft_mount(mountpoint, server, NSFv4_PSEUDOFS, transport, nfsversion='4')
+        util.pread2([NFS_STAT, '-m'])
+        unmount(mountpoint, NSFv4_PSEUDOFS)
+        if res is None:
+            res = ['4']
+        else:
+            res.add('4')
+    except Exception:
+        util.SMlog("Unable to obtain list of valid nfs versions with NSFv4 pseudo FS mount")
```
Is this needed at all when res is not None? IIRC, when the server supports both v3 and v4, rpcinfo already lists 4 among the versions, so there's no need to try to mount the pseudo filesystem from NFSv4 in this situation.

stormi commented on drivers/nfs.py:
```diff
@@ -186,38 +194,74 @@ def unmount(mountpoint, rmmountpoint):
-        entry.appendChild(subentry)
-        textnode = dom.createTextNode(access)
-        subentry.appendChild(textnode)
+    try:
+        mountpoint = "%s/%s" % NFS4_TMP_MOUNTPOINT, target
+        soft_mount(mountpoint, target, NSFv4_PSEUDOFS, transport, nfsversion='4')
+        paths = util.pread2([LS])
+        unmount(mountpoint, NSFv4_PSEUDOFS)
+        for path in paths.split("\n"):
+            entry = dom.createElement('Export')
+            element.appendChild(entry)
+
+            subentry = dom.createElement("Target")
+            entry.appendChild(subentry)
+            textnode = dom.createTextNode(target)
+            subentry.appendChild(textnode)
+            subentry = dom.createElement("Path")
+            entry.appendChild(subentry)
+            textnode = dom.createTextNode(path)
+            subentry.appendChild(textnode)
+
+            subentry = dom.createElement("Accesslist")
+            entry.appendChild(subentry)
+            # Assume everyone as we do not have any info about it
+            textnode = dom.createTextNode("*")
+            subentry.appendChild(textnode)
+        fail = False
+    except Exception:
+        util.SMlog("Unable to scan exports with NFSv4 pseudo FS mount")
```
AFAIK you only need to do this if showmount doesn't give any results. Otherwise, you already have the exports for both v3 and v4.

stormi commented on drivers/nfs.py:
```diff
@@ -186,38 +194,74 @@ def unmount(mountpoint, rmmountpoint):
+            subentry.appendChild(textnode)
+        fail = False
+    except Exception:
+        util.SMlog("Unable to scan exports with rpcingo, trying NFSv4")
```
```suggestion
        util.SMlog("Unable to scan exports with %s, trying NFSv4" % SHOWMOUNT_BIN)
```

stormi commented on drivers/nfs.py:
```diff
@@ -50,34 +52,40 @@
 
+    sv = get_supported_nfs_versions(server, transport)
+    if sv == ['4']:
+        return True
```
Maybe add a comment explaining why?

stormi commented on drivers/nfs.py:
```diff
@@ -186,38 +194,74 @@ def unmount(mountpoint, rmmountpoint):
-        textnode = dom.createTextNode(access)
-        subentry.appendChild(textnode)
+    try:
+        mountpoint = "%s/%s" % NFS4_TMP_MOUNTPOINT, target
```
Does this work without enclosing the tuple in parentheses?

```
>>> a = "%s %s" % "one", "two"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: not enough arguments for format string
```

benjamreis commented:
Successfull tests on NFS3 & 4.

olivierlambert commented:
Hey :) Any news on this? Is there anything left to do or simply getting validated?

MarkSymsCtx commented on drivers/nfs.py:
```diff
@@ -186,43 +182,75 @@ def unmount(mountpoint, rmmountpoint):
-
-
-def scan_srlist(path, dconf):
+    try:
```
I don't particularly like how long this method has now become, it makes it quite cognitively hard to follow. Would be good to pull some of these long open coded blocks out into well named methods which describe what they're doing and then each one can be reviewed and understood in isolation (or ignored if not of interest to the current needs).

MarkSymsCtx replied:
Similarly, to a lesser extent, `get_supported_nfs_versions` below.

benjamreis replied:
I can split in methods for NFS3 and NFS4Only.

benjamreis replied:
Done :)

MarkSymsCtx commented on drivers/nfs.py:
```diff
@@ -218,11 +213,61 @@ def scan_exports(target):
+    return dom
+
+
+def _scan_exports_nfs4_only(target, transport, dom, element):
```
given that we have `_scan_exports_nfs3` this should probably just be `_scan_exports_nfs4` for consistency.

benjamreis replied:
I hesitated to add the `_only` or not, both are okay for me so I can remove it if you prefer.

benjamreis replied:
done :)

MarkSymsCtx reviewed (commented):
Just a minor nit on method names.

Is there anything "useful" to see in the logging for this that will show it working?

benjamreis commented:
> Is there anything "useful" to see in the logging for this that will show it working?

In a NFS3/NFS3-4 env you should see the same logs as before since NFS4 only is a fallback in case of failure.
In NFS4 only env I added logs saying the 1st method fails and we're trying using the NFS4 pseudo FS
https://github.com/xapi-project/sm/pull/617/files#diff-9ca8a7292c88a517d48db9a9d25ee2d14ffd83f7f40d818c1c4121ce2e275b35R258
https://github.com/xapi-project/sm/pull/617/files#diff-9ca8a7292c88a517d48db9a9d25ee2d14ffd83f7f40d818c1c4121ce2e275b35R346

I also added logs in case both methods failed before throwing an exception.

MarkSymsCtx commented on drivers/nfs.py:
```diff
@@ -50,34 +51,41 @@
 
-
-def check_server_tcp(server, nfsversion=DEFAULT_NFSVERSION):
+def check_server_tcp(server, transport, nfsversion=DEFAULT_NFSVERSION):
```
The method is explicit (or is expected to be) in using TCP, so it should not have the transport argument. If `get_suppoted_nfs_versions` needs a transport it should be passed as a literal from this method without changing the signature of the method.

benjamreis replied:
Indeed but transport can be `tcp` or `tcp6` in the case of IPv6.

MarkSymsCtx replied:
Ah, I see, that was not clear. Might have been better for the IPv6 bits to be at least a commit on their own right to make it clear.

MarkSymsCtx commented:
@olivierlambert I'd already added the extra reviewer

olivierlambert commented:
Can you merge? Thanks!

KevinLCtx commented:
> Can you merge? Thanks!

Mark is on vacation I will see if I can find somebody else who has permission to merge it.
